server/hl7/hl7.controller.js

```diff
@@ -108,5 +109,63 @@ function getUserFiles(req, res, next) {
   return next(err);
 }
 
+function getMessageByIndexOrId(req, res, next) {
+  let err;
+  const fileId = req.params.fileId;
+  if (!mongoose.Types.ObjectId.isValid(req.params.indexWithinFileOrId)) { // treat as index in file
+    const messageIndex = req.params.indexWithinFileOrId;
+    if (!Number.isInteger(parseInt(messageIndex, 10))) {
+      err = new APIError(`Invalid Message Index: ${messageIndex}`, httpStatus.BAD_REQUEST);
+      return next(err);
+    }
+    return _getMessageByIndex(messageIndex, fileId)
+      .then((message) => {
+        if (message.length > 0) res.status(httpStatus.OK).json(message);
+        else res.status(httpStatus.NOT_FOUND).json('Message Not found. Check index number & File ID');
+      })
+      .catch((error => next(new APIError(error, httpStatus.INTERNAL_SERVER_ERROR))));
+  }
+  const messageId = req.params.indexWithinFileOrId;
+  return _getMessageById(messageId, fileId)
+    .then((message) => {
+      if (message.length > 0) res.status(httpStatus.OK).json(message);
+      else res.status(httpStatus.NOT_FOUND).json('Message Not found. Check message ID & File ID');
+    })
+    .catch((error => next(new APIError(error, httpStatus.INTERNAL_SERVER_ERROR))));
+}
+
+function _getMessageByIndex(messageIndex, fileId) {
+  return new Promise((resolve, reject) => {
+    Message.find({
```

Comment:
turns out `find` returns an array even if theres only one element. I think we want this to return just a json object for the message rather than an array of one message object

Reply:
there might be a `findOne` or similar method for mongoose

Reply:
`findOne` did it